caf: add support for the BLE conn rate change event (#30156)
nrfconnect/sdk-nrf: doc/nrf/libraries/caf/ble_state.rst

Connection parameter change
===========================

> Connection parameter change is handled differently depending on whether Shorter Connection Intervals (SCI) are enabled or not.

Reviewer: ```suggestion
Connection parameter change is handled differently depending on whether Shorter Connection Intervals (SCI) are user for a given Bluetooth LE connection or not.
```

Reviewer 2: ```suggestion
Connection parameter changes are handled differently depending on whether Shorter Connection Intervals (SCI) are used for a given Bluetooth LE connection.
```

> SCI is controlled by the :kconfig:option:`CONFIG_BT_SHORTER_CONNECTION_INTERVALS` Kconfig option.

Reviewer: ```suggestion
SCI support is controlled by the :kconfig:option:`CONFIG_BT_SHORTER_CONNECTION_INTERVALS` Kconfig option.
```

> Shorter Connection Intervals disabled

Reviewer: used/unused?

Author: changed

-------------------------------------

The module submits a :c:struct:`ble_peer_conn_params_event` to inform other application modules about connection parameter update requests and connection parameter updates.

The |ble_state| rejects the connection parameter update request in Zephyr's callback.
An application module can handle the :c:struct:`ble_peer_conn_params_event` and update the connection parameters.

Shorter Connection Intervals enabled
------------------------------------

> If SCI is enabled, the connection *rate* API (:c:func:`bt_conn_le_conn_rate_request`) should be used to update the connection parameters.

Reviewer: ```suggestion
If SCI is used, the connection *rate* API (:c:func:`bt_conn_le_conn_rate_request`) should be used to update the connection parameters.
```

Reviewer 2: ```suggestion
If SCI is used, use the connection *rate* API (:c:func:`bt_conn_le_conn_rate_request`) to update the connection parameters.
```

The module submits a :c:struct:`ble_peer_conn_rate_event` to inform other application modules about connection rate changes.
The event is also emitted on failed connection rate update requests, containing the status code of the failed request.

> There is no event emitted on the connection rate update request itself, as the Bluetooth stack does not provide a callback for this when SCI is enabled.

Reviewer: Consider informing that BLE Central can use `bt_conn_le_conn_rate_set_defaults` to limit accepted conn rate range.

Author: I see Francesco added it in his suggestion, applied

> There is no event emitted on the connection rate update request itself, as the Bluetooth stack does not provide a callback for this when SCI is enabled.

Reviewer 2: ```suggestion
The event is also emitted when a connection rate update request fails.
In this case, it contains the status code of the failed request.
No event is emitted when a connection rate update request is made, because the Bluetooth stack does not provide a callback for this when SCI is enabled.

A BLE Central can use the :c:func:`bt_conn_le_conn_rate_set_defaults` API to limit the accepted connection rate range.
```

.. note::
   When :kconfig:option:`CONFIG_BT_SHORTER_CONNECTION_INTERVALS` is enabled, the non-SCI callbacks are still available.
   They are called if the non-SCI (connection *parameters* update) API is used to update the connection parameters.

>    However, it is not recommended to use the non-SCI API when SCI is enabled.

Reviewer: Consider mentioning that BLE peer device must also support SCI in order to use those APIs (if peer does not support SCI we cannot use SCI-defined connection parameters and we should stick to "old" APIs, right?)

Author: Done

>    However, it is not recommended to use the non-SCI API when SCI is enabled.

Reviewer 2: ```suggestion
.. note::
   When :kconfig:option:`CONFIG_BT_SHORTER_CONNECTION_INTERVALS` is enabled, the non-SCI callbacks remain available.
   They are called if you use the non-SCI (connection parameter update) API to update the connection parameters.
   However, do not use the non-SCI API when SCI is enabled.
```

Connection parameter change
===========================

Connection parameter changes are handled differently depending on whether Shorter Connection Intervals (SCI) are used for a given Bluetooth LE connection.
SCI support is controlled by the :kconfig:option:`CONFIG_BT_SHORTER_CONNECTION_INTERVALS` Kconfig option.
The peer must also support SCI for it to be used on a connection.

Shorter Connection Intervals unused
-----------------------------------

The module submits a :c:struct:`ble_peer_conn_params_event` to inform other application modules about connection parameter update requests and connection parameter updates.

The |ble_state| rejects the connection parameter update request in Zephyr's callback.
An application module can handle the :c:struct:`ble_peer_conn_params_event` and update the connection parameters.

Shorter Connection Intervals used
---------------------------------

> If SCI is used, use the connection *rate* API (:c:func:`bt_conn_le_conn_rate_request`) to update the connection parameters.

Reviewer 3: ```suggestion
If you are using SCI is, update the connection parameters with the connection *rate* API (:c:func:`bt_conn_le_conn_rate_request`).
```

> If you are using SCI is, update the connection parameters with the connection *rate* API (:c:func:`bt_conn_le_conn_rate_request`).

Reviewer 4: ```suggestion
If you are using SCI, update the connection parameters with the connection *rate* API (:c:func:`bt_conn_le_conn_rate_request`).
```

The module submits a :c:struct:`ble_peer_sci_conn_rate_event` to inform other application modules about connection rate changes.
The event is also emitted when a connection rate update request fails.
In this case, it contains the status code of the failed request.
No event is emitted when a connection rate update request is made, because the Bluetooth stack does not provide a callback for this when SCI is enabled.

> A BLE Central can use the :c:func:`bt_conn_le_conn_rate_set_defaults` API to limit the accepted connection rate range.

Reviewer 3: ```suggestion
A Bluetooth LE Central can use the :c:func:`bt_conn_le_conn_rate_set_defaults` API to limit the accepted connection rate range.
```

.. note::

>    When :kconfig:option:`CONFIG_BT_SHORTER_CONNECTION_INTERVALS` is enabled, the non-SCI callbacks remain available.

Reviewer 3: ```suggestion
   When the :kconfig:option:`CONFIG_BT_SHORTER_CONNECTION_INTERVALS` Kconfig option is enabled, the non-SCI callbacks remain available.
```

   They are called if you use the non-SCI (connection parameter update) API to update the connection parameters.

>    However, do not use the non-SCI API when SCI is enabled.

Reviewer: ```suggestion
   However, do not use the non-SCI API if both the device and peer support the SCI.
```
